### PR TITLE
feat: `DistributionManager`

### PR DESCRIPTION
**Pull request by deluca-mike**
- `PayerReportManager.getPayerReports` changed to query an array of originator nodeIds and payer report indices

## Summary by CodeRabbit

- **New Features**
  - Introduced a contract for managing and distributing fees to node owners, including claiming and withdrawing owed fees.
  - Added interfaces and events to support fee claim tracking and secure withdrawals.
  - Provided migration support for future upgrades.

- **Enhancements**
  - Updated report retrieval functions to support batch queries and improved error handling for mismatched input arrays.
  - Extended interfaces to improve compatibility and ownership verification.

- **Bug Fixes**
  - Improved test coverage for error cases and state transitions in fee distribution and report management.

- **Tests**
  - Added comprehensive tests for fee claiming, withdrawal, migration, and edge cases.
  - Introduced new test harnesses and mock contracts to facilitate robust testing.

- **Chores**
  - Updated configuration by removing an unused linting rule.

- **CI**
  - Added a fixed fuzzing seed to improve determinism in gas report testing.

### CODE REVIEW

File: test/utils/Mocks.sol [hunk 1]
```diff
@@ -95,3 +96,20 @@ contract MockPayerRegistry {
...
+        bool isSettled;
+        bytes32 payersMerkleRoot;
+        uint32[] nodeIds;
+    }
+
+    function getPayerReports(
+        uint32[] calldata originatorNodeIds_,
+        uint256[] calldata payerReportIndices_
+    ) external view returns (PayerReport[] memory payerReports_) {}
+}
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply test/utils/Mocks.sol:115]: This is intended behaviour. Tests that will result in this function being called should always mock it. If they forget to, they will get an error.

File: src/settlement-chain/DistributionManager.sol [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+        DistributionManagerStorage storage $ = _getDistributionManagerStorage();
+
+        for (uint256 index_; index_ < payerReports_.length; ++index_) {
+            // This makes the assumption that the returned array of payer reports from `getPayerReports` is of the same
+            // length, and in the same order, as the arrays of originator node IDs and payer report indices.
+            uint32 originatorNodeId_ = originatorNodeIds_[index_];
+            uint256 payerReportIndex_ = payerReportIndices_[index_];
+
+            if ($.hasClaimed[nodeId_][originatorNodeId_][payerReportIndex_]) {
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply src/settlement-chain/DistributionManager.sol:108]: The `DistributionManager` and `PayerReportManager` contracts are all first party contracts, so their behaviour is largely expected. Integration tests and simulations will always be performed to ensure correct integration and to catch these issues when implementations change.

File: src/settlement-chain/DistributionManager.sol [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            }
+
+            $.hasClaimed[nodeId_][originatorNodeId_][payerReportIndex_] = true;
+
+            // `nodeIds_.length` must be at least 1 for `_isInNodeIds` to have returned `true`.
+            uint96 feePortion_ = payerReport_.feesSettled / uint96(nodeIds_.length);
+
+            unchecked {
+                claimed_ += feePortion_;
+            }
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply src/settlement-chain/DistributionManager.sol:129]: We're operating under the assumption that the underlying token being used for fee has a total supply well under `type(uint96).max`. The contract will be upgraded or modified if that ceases to be the case.

File: src/settlement-chain/DistributionManager.sol [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    function claim(
+        uint32 nodeId_,
+        uint32[] calldata originatorNodeIds_,
+        uint256[] calldata payerReportIndices_
+    ) external returns (uint96 claimed_) {
+        _revertIfNotNodeOwner(nodeId_);
+
+        IPayerReportManagerLike.PayerReport[] memory payerReports_ = IPayerReportManagerLike(payerReportManager)
+            .getPayerReports(originatorNodeIds_, payerReportIndices_);
+
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply src/settlement-chain/DistributionManager.sol:99]: While correct, we rely on the `getPayerReports` function at the payer report manager to revert if this is the case. However, it is a cheap enough check that we can add this check to distribution manager's claim function ss well, just to separate concerns and increase robustness.

File: src/settlement-chain/DistributionManager.sol [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        if (availableBalance_ == 0) revert ZeroAvailableBalance();
+
+        // Only withdraw up to what is available.
+        withdrawn_ = availableBalance_ >= owedFees_ ? owedFees_ : availableBalance_;
+
+        unchecked {
+            // `withdrawn_` is less than or equal to `owedFees_`, and `totalOwedFees` is the sum of all `owedFees`.
+            $.owedFees[nodeId_] = owedFees_ - withdrawn_;
+            $.totalOwedFees -= withdrawn_;
+        }
```
Comment (fbac) [inline src/settlement-chain/DistributionManager.sol:164]: nit: code is good, but I'm wondering if as a Node Operator I'd expect an all-or-nothing approach, instead of calling withdraw N times over a couple days and spending some extra gas.

What is clear is that it's on us calling the send excess function via an automated off-chain service, to make the UX as good as possible.

**Review by fbac — APPROVED** [review]

File: src/settlement-chain/DistributionManager.sol (hunk 5 above)
Comment (deluca-mike) [reply src/settlement-chain/DistributionManager.sol:164]: The process is:
- incrementally settle payer reports, which sets aside payer balances and creates debt if needed
- send excess funds from the payer registry to the distribution manager
- node owners can claim fee for any fully settled payer reports, and increase their owed balances
- whenever funds become available, or if they already are available, node owners can withdraw

The idea is that these can all be done asynchronously so that they are not coupled and to minimize the permutations of expected outcomes.

The reason the sending of excess is separate is because:
- the gas associated with it will only be required once per batch of incremental payer report settlements that may happen in rapid succession within aa short time frame
- new funds may become available as indebted payers deposit more funds

However, as funds are only really needed when node owners withdraw, we can add to the `withdraw` function the calling of `payerRegistry.sendExcessToFeeDistributor` if there are insufficient funds for the withdrawal. That will cut down on the need for any automated off-chain service.

File: test/unit/GroupMessageBroadcaster.t.sol [hunk 6]
```diff
@@ -152,10 +152,10 @@ contract GroupMessageBroadcasterTests is Test {
...
         bool paused_
     ) external {
-        minPayloadSize_ = bound(minPayloadSize_, 1, _STARTING_MAX_PAYLOAD_SIZE);
-        maxPayloadSize_ = bound(maxPayloadSize_, minPayloadSize_, _STARTING_MAX_PAYLOAD_SIZE);
-        payloadSize_ = bound(payloadSize_, 1, _STARTING_MAX_PAYLOAD_SIZE);
-        sequenceId_ = uint64(bound(sequenceId_, 0, type(uint64).max - 1));
+        minPayloadSize_ = _bound(minPayloadSize_, 1, _STARTING_MAX_PAYLOAD_SIZE);
+        maxPayloadSize_ = _bound(maxPayloadSize_, minPayloadSize_, _STARTING_MAX_PAYLOAD_SIZE);
+        payloadSize_ = _bound(payloadSize_, 1, _STARTING_MAX_PAYLOAD_SIZE);
+        sequenceId_ = uint64(_bound(sequenceId_, 0, type(uint64).max - 1));
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply test/unit/GroupMessageBroadcaster.t.sol:158]: `_bound` is inherited from `Test` from the `forge-std` import. The `bound` inherited from `Test` has an odd side-effect where it console logs all the values that pass through it, which slow down fuzz tests.
